src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -623,95 +636,76 @@ private static void CloneAndAddValue(HeaderStoreItemInfo destinationInfo, object
             }
         }
 
-        private HeaderStoreItemInfo GetOrCreateHeaderInfo(HeaderDescriptor descriptor, bool parseRawValues)
+        private HeaderStoreItemInfo GetOrCreateHeaderInfo(HeaderDescriptor descriptor)
         {
-            HeaderStoreItemInfo? result = null;
-            bool found;
-            if (parseRawValues)
+            if (TryGetAndParseHeaderInfo(descriptor, out HeaderStoreItemInfo? info))
```

Comment:
This seems like it could benefit from the GetValueRefOrAddDefault optimization.

Reply:
We sort-of already do in the common case. In the `else` branch we call `AddEntryToStore`, which asserts that the key doesn't exist yet and doesn't perform the whole scan.

Getting back a `ref` to use here would be a bigger change since `TryGetAndParseHeaderInfo` will not only parse, but also remove the entry if it's not present.